core/src/Kokkos_HPX.hpp

```diff
@@ -191,21 +140,26 @@ class HPX {
 
   struct instance_data {
     instance_data() = default;
-    instance_data(hpx::shared_future<void> future) : m_future(future) {}
-    Kokkos::Impl::thread_buffer m_buffer;
-    hpx::shared_future<void> m_future = hpx::make_ready_future<void>();
-    hpx::spinlock m_future_mutex;
+    instance_data(uint32_t instance_id) : m_instance_id(instance_id) {}
+    instance_data(uint32_t instance_id,
+                  hpx::execution::experimental::unique_any_sender<> &&sender)
+        : m_instance_id(instance_id), m_sender{std::move(sender)} {}
+
+    instance_data(const instance_data &) = delete;
+    instance_data(instance_data &&)      = delete;
+    instance_data &operator=(const instance_data &) = delete;
+    instance_data &operator=(instance_data) = delete;
+
+    uint32_t m_instance_id{HPX::impl_default_instance_id()};
+    hpx::execution::experimental::unique_any_sender<> m_sender{
+        hpx::execution::experimental::just()};
+    Kokkos::Impl::hpx_thread_buffer m_buffer;
+    hpx::spinlock m_sender_mutex;
   };
 
-  mutable std::shared_ptr<instance_data> m_independent_instance_data;
+  static void default_instance_deleter(instance_data *) {}
   static instance_data m_default_instance_data;
-
-  std::reference_wrapper<Kokkos::Impl::thread_buffer> m_buffer;
-  std::reference_wrapper<hpx::shared_future<void>> m_future;
-  std::reference_wrapper<hpx::spinlock> m_future_mutex;
-#else
-  static Kokkos::Impl::thread_buffer m_default_buffer;
-#endif
+  std::shared_ptr<instance_data> m_instance_data;
```

Comment:
We typically have been using `Impl::HostSharedPtr` instead of `std::shared_ptr` for the pointer to implementation in other backends.
If I recall properly it was to allow execution space instances to be data member in user-defined classes that get captured onto the device (not that we recommend doing that...)

Reply:
Yes

Reply:
Changed in https://github.com/kokkos/kokkos/pull/5628/commits/6976a03128324d4318d1a3698c351e6b8daac9bc.